### DISCUSSION TITLE
How to properly configure webpack to build React Hook Form

### DISCUSSION THREAD

**Discussion by brtchkv** [react-hook-form/react-hook-form#2387] [Q&A] [answered]
I'm constantly getting an error in an attempt to build TypeScript ([example project](https://codesandbox.io/embed/react-hook-form-v6-ts-errormessage-multiple-error-messages-jk43e?file=/src/index.tsx)) from the documentation. 

Webpack keeps telling me that `TS2558: Expected 0 type arguments, but got 1.` on the line where `useFormuseForm<IFormInputs>({ criteriaMode: "all"});` is declared
I'm using
```
"react-hook-form": "^6.1.0"
"ts-loader": "^8.0.1",
"typescript": "^3.9.7",
"webpack": "^4.44.0",
```

UPD: adding @tsignore annotation did help, however I can't fix errors that happen when web pack is trying to resolve error attributes (e.g errors.username)

**Comment by ikibalnyi** [comment]
`TS2558: Expected 0 type arguments, but got 1. ` basically means that you are trying to pass a type argument to a function/interface which is not generic.
For example:
```
interface Test {}
const fn = () => {};

fn<Test>(); // this will throw with TS2558: Expected 0 type arguments, but got 1.
```
Make sure you run `yarn`/`npm i`.
Make sure you use `useForm` from `react-hook-form`;
If it doesn't help, could you please paste here the whole line.
Also providing webpack config and tsconfig would be helpful as from example it's not clear where this error can come from. 
Also does your text editor show you errors? or does it happen only during webpack build/watch?

**Reply by brtchkv** [reply]
Hi, 

tsconfig:
```
{
  "compilerOptions": {
    "outDir": "./dist/",
    "noImplicitAny": false,
    "module": "es6",
    "target": "es5",
    "jsx": "react",
    "allowJs": true,
    "allowSyntheticDefaultImports": true
  }
}
```

webpack.config
```
var glob = require("glob");

const { resolve } = require('path')
const webpack = require('webpack')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const OfflinePlugin = require('offline-plugin')
const PreloadWebpackPlugin = require('preload-webpack-plugin')
const LoadablePlugin = require('@loadable/webpack-plugin')

module.exports = {
    entry: {
        main: './src/index.js'
    },
    output: {
        // `filename` provides a template for naming your bundles (remember to use `[name]`)
        filename: '[name].bundle.js',
        // `chunkFilename` provides a template for naming code-split bundles (optional)
        chunkFilename: '[name].bundle.js',
        // `path` is the folder where Webpack will place your bundles
        path: '/home/ivan/Documents/plp/build',
        // `publicPath` is where Webpack will load your bundles from (optional)
        publicPath: 'dist/'
    },
    module: {
        rules: [
            {
                test: /\.jsx?$/,
                exclude: /(node_modules|bower_components)/,
                use: {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-react"]
                    }
                }
            },
            {
                test: /\.css$/,
                use: [
                    {
                        loader: "css-loader"
                    }
                ]
            },
            {
                test: /\.(png|gif|woff|woff2|eot|ttf|svg)$/,
                loader: "url-loader?limit=100000"
            },
            {
                test: /\.s[ac]ss$/i,
                use: [
                    // Creates `style` nodes from JS strings
                    'style-loader',
                    // Translates CSS into CommonJS
                    'css-loader',
                    // Compiles Sass to CSS
                    'sass-loader',
                ],
            },
            {
                test: /\.tsx?$/,
                use: 'ts-loader',
                exclude: /node_modules/,
            },
        ],
    },
    resolve: {
        extensions: [ '.tsx', '.ts', '.js' ],
    },
};
```

Also here's the whole code of the component: ([pastebin](https://pastebin.com/HU7JkURs))

Errors, which I get:
<img width="660" alt="image" src="https://user-images.githubusercontent.com/17526135/88599503-9868a500-d074-11ea-8ea4-6080742ee932.png">

**Comment by ikibalnyi** [comment, marked as answer]
looks like your tsconfig is missing `"moduleResolution": "node"` under `compilerOptions`, so typescript is not able to resolve absolute imports properly https://www.typescriptlang.org/docs/handbook/module-resolution.html

**Reply by ikibalnyi** [reply]
Also `schema.properties.password.required` should be `boolean`

**Reply by brtchkv** [reply]
omg, thank u so much, moduleResolution fixed all of the errors!
